src/core/fair_queue.cc

```diff
@@ -159,6 +159,14 @@ class fair_queue::priority_class_data {
     void update_shares(uint32_t shares) noexcept {
         _shares = (std::max(shares, 1u));
     }
+
+    bool has_noncancelled_entry() const {
+        bool result = false;
+        for (auto& ent : _queue) {
+            result |= ent._capacity > 0;
```

Comment:
There's an opinion that `|=` doesn't work for bool types:
https://github.com/scylladb/seastar/pull/1841#discussion_r1331386016
https://stackoverflow.com/questions/2488406/why-doesnt-c-have-or-for-booleans

Reply:
I think |= works as long as everything is bool, no?

> A bool may only be true or false in C++. As such, using &= and |= is relatively safe (even though I don’t particularly like the notation). True, they will perform bit operations rather than logical operations (and thus they won’t short-circuit) but these bit operations follow a well-defined mapping, which is effectively equivalent to the logical operations, as long as both operands are of type bool.1

Reply:
I don't know:

> "&=" is not the right operator, it's the bitwise operator :-(
> I know some people claim it works (https://stackoverflow.com/questions/2488406/why-doesnt-c-have-or-for-booleans) but I think it's wrong.

For me it works, but Nadav (in comment) correctly says that |= is not logical, but a bitwise operator. The correct one here should be ||=, but there's no such in C++

Reply:
there's a big difference between &= and |=. &= won't work for bool_var &= 3 but |= always works.

Reply:
I don't see the problem even with &= in the first place. Compiler knows its bool values and can handle it correctly. However, if we try to find interpretation that leads to broken behavior then even for `|= 3` things can go wrong
```
bool x = true; // x = 0x1
x |= 3; // x = 0x3
x == true; // 0x3 == 0x1 -> can be either true or false
```

Reply:
So I'll leave it up to our compilers expert (@tchaikov ). Kefu, do you think the problem is invented out of nowhere or should indeed avoid |= here?

Reply:
I don't think `x == true` can ever be false if x is true. No matter what the underlying value of x is. This is bool comparison so all values have to be casted to bool otherwise this wouldn't work either:
```
bool a = 1;
bool b = 2;
assert(a == b);
```

Reply:
if i were a C++17 compiler implemented using static single assignment, and am requested to compile following source list:
```c++
bool x = true;
x |= 3;
x == true;
```
i would first translate it to something like
```c++
bool x = true;
int x1 = static_cast<int>(x); // x1 = 0x1, as x should be promoted. so that it can participate as the LHS of `|` operation.
int x2 = x1|3; // x2 = 0x1 | 0x3 = 3
bool x3 = static_cast<bool>(x2); // x3 = true. as it casted back to the type of the LHS
bool y = x3 == true; // x3 == true
```

see also https://eel.is/c++draft/expr.ass#6 and https://eel.is/c++draft/expr#conv.bool

probably we can just sum up the `ent._capacity` so we can avoid the branching in the tight loop, if the sum  does not overflow.

Reply:
> I don't think `x == true` can ever be false if x is true. No matter what the underlying value of x is. This is bool comparison so all values have to be casted to bool otherwise this wouldn't work either:
> 
> ```
> bool a = 1;
> bool b = 2;
> assert(a == b);
> ```

yeah, you beat me to it. i concur with your point. just one thing the "underlying value of x", from the compiler's point of view should always be `true` or `false`. but from the machine's perspective, it's always 1 or 0.